server/lib/adapters/mssql.js

```diff
@@ -0,0 +1,231 @@
+      };
+    }
+    return false;
+  }, {});
+  // Must-have options
+  let opts = {
+    port: conn.port || 1433,
+    database: conn.name
+  };
+  // All additional options are optional
+  if (settings.db_options) {
+    try {
+      const customOptions = JSON.parse(settings.db_options);
+      if (customOptions) {
+        opts = _.merge(opts, customOptions);
+      }
+    } catch (parseError) {
+      console.error("config.error", parseError);
+    }
+  }
+
+  const config = {
+    userName: conn.user,
+    password: conn.password,
+    server: conn.host,
+    options: opts
+  };
+
+  return config;
+}
+/**
+ * Open a new connection.
+ *
+ * @param {Object} settings The ship settings.
+ *
+ * @return {tedious.Connection} The tedious.Connection instance
+ */
+export function openConnection(settings) {
+  const config = parseConnectionConfig(settings);
+  return new tedious.Connection(config);
+}
+
+/**
+ * Close the connection.
+ *
+ * @param {tedious.Connection} client The MSSQL client
+ */
+export function closeConnection(client) {
+  client.close();
+}
+
+/**
+ * Wrap the user query inside a MS SQL request.
+ *
+ * @param {*} sql The raw SQL query
+ * @param {*} replacements The replacement parameters
+ */
+export function wrapQuery(sql, replacements) {
+  return SequelizeUtils.formatNamedParameters(sql, replacements, "mssql");
+}
+
+/**
+ * Runs the query using the specified client and options.
+ * @param {tedious.Connection} client The MS SQL client.
+ * @param {string} query The query to execute.
+ * @param {Object} options The options.
+ *
+ * @returns {Promise} A promise object of the following format: { rows }
+ */
+export function runQuery(client, query, options) {
+  return new Promise((resolve, reject) => {
+    const confoptions = _.merge(client.config.options, options);
+    const conf = _.cloneDeep(client.config);
+    conf.options = confoptions;
+
+    const conn = new tedious.Connection(conf);
+
+    conn.on("connect", (err) => { // eslint-disable-line consistent-return
+      if (err) {
+        return reject(err);
+      }
+
+      const qparam = `WITH __qry__ AS (${query}) SELECT TOP(100) * FROM __qry__`;
+
+      const request = new tedious.Request(qparam, (reqError) => { // eslint-disable-line consistent-return
+        if (reqError) {
+          return reject(reqError);
+        }
+      });
+
+      const rows = [];
+      let resultReturned = false;
+
+      request.on("row", (columns) => {
+        const row = {};
+        _.forEach(columns, (column) => {
+          row[column.metadata.colName] = column.value;
+        });
+        rows.push(row);
+      });
+
+      request.on("done", (rowCount) => { // eslint-disable-line consistent-return
+        if (_.isNumber(rowCount) && !resultReturned) {
+          resultReturned = true;
+          return resolve({
+            rows
+          });
+        }
+      });
+
+      request.on("doneInProc", (rowCount) => { // eslint-disable-line consistent-return
```

Comment:
possibly we could have a small handler function to pass in case of each event just to clearly mark that each event is basically the same in terms of logic flow we want there.